### ISSUE TITLE
Missing edge case information in BCELoss documentation

### ISSUE THREAD

**Issue by Enealor** [pytorch/pytorch#31453]
## 📚 Documentation

The documentation for [BCELoss](https://pytorch.org/docs/stable/nn.html#bceloss) explains that the formula used to compute the loss is `l_n = -w_n [y_n \ln(x_n)+(1-y_n)\ln(1-x_n)]`. However, the documentation does not explain what happens when `x_n = 0, 1` with `0 < y_n < 1`. In these two cases, the result should be infinite.

It appears that the value is quietly substituted for stability. However, there is no mention of this in the documentation for the loss layer, nor is it explained how this is accomplished.

I also could not find the code that implemented binary_cross entropy. I bring this up because the code should be consulted when explaining what is happening for these edge cases.

(This came up while investigating #31396)

cc @ezyang @gchanan @zou3519

**Comment by albanD** [comment]
Setting high priority as it should be fairly easy to fix.

**Comment by aayn** [comment]
Can I take this up (new contributor here)? From preliminary exploration, it seems like when computing `log(0)` in this case, PyTorch internally computes `log(1e-12)` as can be seen [here](https://github.com/pytorch/pytorch/blob/3397d41b8af2824495faaefc649c3c7e47f03bfb/aten/src/THNN/generic/BCECriterion.c#L5) and [here](https://github.com/pytorch/pytorch/blob/ffe0c1ae4d8de75bc4d6cf5c43123df1e538d8a4/aten/src/THCUNN/BCECriterion.cu#L26). This can be verified as follows:
![bce_pytorch](https://user-images.githubusercontent.com/9693267/71325431-80f38900-24a1-11ea-9315-c0886ec6969a.png)

**Comment by Enealor** [comment]
Thanks for the find. I am still getting used to navigating the C/C++ code. I can see now that it is using a safe log function that replaces `log(0)` specifically with `log(EPS)` where `EPS = 1e-12`. If you would make a pull request to update the documentation, that would be great.

**Comment by aayn** [comment]
I'll do that! This brought something interesting to my attention. The binary cross entropy loss for values less than `1e-12` but greater than `0` is greater than the loss for `0`. For example:
![maybe_prob](https://user-images.githubusercontent.com/9693267/71325955-1fcfb380-24a9-11ea-967a-86ffca971047.png)

Is this behavior desirable/acceptable or should it be fixed?

**Comment by kurtamohler** [comment]
FYI, I recently created a PR to port BCELoss from TH to ATen (https://github.com/pytorch/pytorch/pull/31365), and I expect it will be merged in the near future. I could take over this issue and update the documentation to match it. Or, @aayn, if you would still like to do it, then I would recommend waiting to push your documentation changes until after my changes have been merged.

As already mentioned above, the bce loss equation gives and infinite result if x = 0 or 1. Infinite results aren't desirable, especially considering that an infinite gradient would mess up training methods. So the current implementation gets avoids infinity by checking if the input to the log() function is exactly zero, and if so, the result becomes log(1e-12) instead (in file aten/src/THNN/generic/BCECriterion.c). There's a bit of an issue with that though--a float can be less than 1e-12 and greater than 0.

Furthermore, this way of avoiding infinity ends up giving a different result than BCEWithLogitsLoss for these edge cases. So in my PR, I changed it to match BCEWithLogitsLoss. I clamped the output of the log() function calls to a minimum of -100 (which is equivalent to clamping the input to e^(-100) = 3.720076e-44). You can look at the file aten/src/ATen/native/Loss.cpp in my PR to see that.

**Comment by aayn** [comment]
Your clamped log update makes sense. You can take over the issue @kurtamohler.
